metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,395 @@
+                         nb.backward[i].kmer_orientation);
+            }
+        }
+    }, dict_);
+}
+template <>
+void DBGSSHash::call_incoming_kmers_with_rc<true>(
+        node_index,
+        const std::function<void(node_index, char, bool)>&) const;
+template <>
+void DBGSSHash::call_incoming_kmers_with_rc<false>(
+        node_index,
+        const std::function<void(node_index, char, bool)>&) const;
+
+void DBGSSHash::call_outgoing_kmers(node_index node,
+                                    const OutgoingEdgeCallback& callback) const {
+    if (mode_ == CANONICAL) {
+        call_outgoing_kmers_with_rc<true>(node, [&](node_index next, char c, bool orientation) {
+            callback(orientation ? reverse_complement(next) : next, c);
+        });
+    } else {
+        call_outgoing_kmers_with_rc<false>(node, [&](node_index next, char c, bool) {
+            callback(next, c);
+        });
+    }
+}
+
+void DBGSSHash::call_incoming_kmers(node_index node,
+                                    const IncomingEdgeCallback& callback) const {
+    if (mode_ == CANONICAL) {
+        call_incoming_kmers_with_rc<true>(node, [&](node_index prev, char c, bool orientation) {
+            callback(orientation ? reverse_complement(prev) : prev, c);
+        });
+    } else {
+        call_incoming_kmers_with_rc<false>(node, [&](node_index prev, char c, bool) {
+            callback(prev, c);
+        });
+    }
+}
+
+size_t DBGSSHash::outdegree(node_index node) const {
+    assert(node > 0 && node <= num_nodes());
+    size_t res = 0;
+    adjacent_outgoing_nodes(node, [&](node_index) { ++res; });
+    return res;
+}
+
+size_t DBGSSHash::indegree(node_index node) const {
+    assert(node > 0 && node <= num_nodes());
+    size_t res = 0;
+    adjacent_incoming_nodes(node, [&](node_index) { ++res; });
+    return res;
+}
+
+void DBGSSHash::call_nodes(
+        const std::function<void(node_index)>& callback,
+        const std::function<bool()> &terminate) const {
+    for (size_t node_idx = 1; !terminate() && node_idx <= dict_size(); ++node_idx) {
+        callback(node_idx);
+        if (mode_ == CANONICAL && !terminate()) {
+            size_t rc_node_idx = reverse_complement(node_idx);
+            if (rc_node_idx != node_idx)
+                callback(rc_node_idx);
+        }
+    }
+}
+
+template <bool with_rc>
+std::pair<DBGSSHash::node_index, bool>
+DBGSSHash::kmer_to_node_with_rc(std::string_view kmer) const {
+    if (!num_nodes())
+        return std::make_pair(npos, false);
+
+    return std::visit([&](const auto &d) {
+        auto res = d.lookup_advanced(kmer.data(), with_rc);
+        return std::make_pair(sshash_to_graph_index(res.kmer_id), res.kmer_orientation);
+    }, dict_);
+}
+template <>
+std::pair<DBGSSHash::node_index, bool>
+DBGSSHash::kmer_to_node_with_rc<true>(std::string_view) const;
+template <>
+std::pair<DBGSSHash::node_index, bool>
+DBGSSHash::kmer_to_node_with_rc<false>(std::string_view) const;
```

Comment:
Huh, why? Isn't it enough to have `template<bool with_rc>` version above?

Reply:
Oh, is it to fix linking errors? You probably want to move it to the header then, I assume?

Reply:
Hmm...so it turned out that rearranging the functions in the cpp fixed things. The rule is that one should provide the implementations of the template methods before the implementations of any other methods that instantiate them.

Reply:
> Oh, is it to fix linking errors? You probably want to move it to the header then, I assume?

Given that the implementation of `kmer_to_node_with_rc` uses some implementation details of the class (e.g., the node index shifting), I'd rather keep it in the cpp if possible.

Reply:
> Hmm...so it turned out that rearranging the functions in the cpp fixed things. The rule is that one should provide the implementations of the template methods before the implementations of any other methods that instantiate them.

turns out it did need explicit instantiation, I was just doing it wrong...